### PR TITLE
collision resolution

### PR DESCRIPTION
**Pull request by mikhail-vlasenko**
attempts #16 

full (partially) elastic collisions for arbitrary shapes seem pretty difficult. so I started small for now

### CODE REVIEW

File: cotix/_bodies.py [hunk 1]
```diff
@@ -15,11 +15,13 @@ class AbstractBody(eqx.Module, strict=True):
...
 
-    position: AbstractVar[Float[Array, "2"]]
+    position: AbstractVar[
+        Float[Array, "2"]
+    ]  # having position here and in shape is inconvenient
     velocity: AbstractVar[Float[Array, "2"]]
 
     angle: AbstractVar[Float[Array, ""]]
-    angular_velocity: AbstractVar[Float[Array, "2"]]
+    angular_velocity: AbstractVar[Float[Array, "2"]]  # shouldnt this be a scalar?
```
Comment (knyazer) [inline cotix/_bodies.py:24]: Just finished ramming jaxtyping through the codebase, and this one one of two mistakes :)

File: cotix/_collision_resolution.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    # lets apply translation to both bodies, taking their mass into account
+    epsilon = 1e-3
+    body1 = body1.set_position(
+        body1.position
+        + (1 + epsilon) * epa_vector * (body2.mass / (body1.mass + body2.mass))
+    )
+    body2 = body2.set_position(
+        body2.position
+        - (1 + epsilon) * epa_vector * (body1.mass / (body1.mass + body2.mass))
+    )
```
Comment (knyazer) [inline cotix/_collision_resolution.py:20]: I think we need the resolution with smaller size or equal size than EPA. Like, we are alright with bodies being a bit inside each other, but not alright if they start oscillate like crazy. The second scenario might happen when you have ```1+epsilon```

Comment (mikhail-vlasenko) [reply cotix/_collision_resolution.py:20]: tbh i added epsilon only when the tests weren't always passing. lets discuss tmrw in-depth, I have quite some questions

**Comment by knyazer** [comment]
Also, why only balls? The collision resolution algorithm that you have should work almost perfectly with anything else. Though computing surface normal seems hard, in practice it should not be hard: we should just use ```jax.grad``` on the support function. This is actually such a smart move, that I think we could make it a really small paper, or blogpost, just about using autodifferention in this particular fashion. Like, usually computing normals to arbitrary shapes is a complete pain

**Comment by knyazer** [comment]
Though it might be harder than I expect, since the surface tangent is not precisely a subgradient: like, we have two bodies, and only one of the subgradients is going to be an appropriate surface tangent... Cuz, if you imaging like a polygon colliding with a circle, the surface normal will be the normal of the circle, not the normal of the polygon. So we will probably have to think about it harder.

Maybe randomized smoothing? But I am not sure...

**Comment by knyazer** [comment]
When I think about computing normal correctly, another obvious thing that comes to mind is just using the penetration vector direction as the normal: since the shapes are convex, the "shortest path to the surface" is going to be normal to the surface, and EPA effectively computes the shortest path to the surface in minkowski difference space. But, I am not precisely sure that being normal in minkowski difference space means being normal in normal space: it is going to be something in the lines of being normal 'simultaneously to both bodies' which is weird. Idk..

**Comment by mikhail-vlasenko** [comment]
im going to extend the ball resolution to any shape, treating the EPA vector as the normal

File: cotix/_collision_resolution.py (hunk 2 above)
Comment (mikhail-vlasenko) [reply cotix/_collision_resolution.py:20]: alright, will multiply by `1/5` or something rather than `(1 + epsilon)`

Comment (mikhail-vlasenko) [reply cotix/_collision_resolution.py:20]: i thought again about the 1/5, and the walls example, why cant we just make the walls immovable? and not update their position. with the gradual position change, we will also want a gradual velocity change, right? or do we just let the bodies flow out of each other after the initial speed change?
can still remove the epsilon though

**Comment by mikhail-vlasenko** [comment]
so this doesnt consider angular momentum yet, but I would prefer to merge it, and then have another PR for angular momentum, as there are still some things missing for it: like center of mass, or the function for produce mass matrix of the body, which I would like to put into the body class.

**Review by knyazer — COMMENTED** [review]
LGTM! Thanks

**Review by knyazer — APPROVED** [review]
Argh, GitHub on mobile is a pain. Here is an approve...

File: cotix/_bodies.py [hunk 3]
```diff
@@ -107,12 +109,29 @@ class Ball(AbstractBody, strict=True):
...
+        ball = (
+            ball.set_mass(1.0)
+            .set_inertia(1.0)
+            .set_position(jnp.zeros((2,)))
+            .set_velocity(jnp.zeros((2,)))
+            .set_angle(0)
+            .set_angular_velocity(0)
+            .set_shape(Circle(0.05, jnp.zeros((2,))))
+        )
+        return ball
```
[the remark this answers is not in the document]
Comment (mikhail-vlasenko) [reply cotix/_bodies.py]: oh i got it, you wanted to basically have it like a singleton. yeah I wanted to use balls for tests of collisions so I changed that

File: cotix/_collision_resolution.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    v1q_new, v2q_new = _1d_elastic_collision_velocities(
+        body1.mass, body2.mass, v1q, v2q
+    )
+
+    v1qr_new = jnp.array([v1q_new * elasticity, v1r])
+    v2qr_new = jnp.array([v2q_new * elasticity, v2r])
+
+    v1_new = jnp.matmul(change_of_basis_inv, v1qr_new)
+    v2_new = jnp.matmul(change_of_basis_inv, v2qr_new)
```
[the remark this answers is not in the document]
Comment (mikhail-vlasenko) [reply cotix/_collision_resolution.py:61]: yeah im using it in numpy all the time, but wasn't completely sure jax properly handles that so went with a safer option lol

File: cotix/_collision_resolution.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    body1: Ball, body2: Ball, epa_vector: Float[Array, "2"]
+) -> (Ball, Ball):
+    elasticity = 1
+
+    # change coordinate system from (x, y) to (q, r)
+    # where q is the line between the centers of the balls
+    unit_collision_vector = epa_vector / jnp.linalg.norm(epa_vector)
+    perpendicular = perpendicular_vector(unit_collision_vector)
+    change_of_basis = jnp.array([unit_collision_vector, perpendicular])
+    change_of_basis_inv = jnp.linalg.inv(change_of_basis)
```
[the remark this answers is not in the document]
Comment (mikhail-vlasenko) [reply cotix/_collision_resolution.py:46]: for any orthonormal matrix, the inverse exists, which is exactly what we are dealing with here, granted  jnp.linalg.norm(epa_vector) != 0

**Pull request merged**